crypto/rand/rand_unix.c

```diff
@@ -203,7 +204,8 @@ size_t rand_pool_acquire_entropy(RAND_POOL *pool)
                 if (fread(buffer, 1, bytes_needed, fp) == bytes_needed)
                     bytes = bytes_needed;
 
-                entropy_available = rand_pool_add_end(pool, bytes, 8 * bytes);
+                rand_pool_add_end(pool, bytes, 8 * bytes);
+                entropy_available = rand_pool_entropy_available(pool);
```

Comment:
I remember that at the time I wrote the api, the return value was orignally a boolean. But then I followed the useful advice of somebody whose name I've forgotten:

> If you have the choice whether just to return a boolean indication of success, or something useful in the case of success, then return something useful.

So the boolean return value in our case indicates whether the function was able to fill up the pool with the required amount entropy or not. The caller will only see 'yes' or 'no'. If however I return the entropy amount in the success case and 0 otherwise, then the caller not only has an indication of success but also something reusable, the fill level of the entropy pool, in the case of success. 

The visible indication of the usefullness of this principle is the fact that all your two-liners are one-liners in my version. And I don't loose any information, because the error state and the i'm-not-ready-yet state are easily distinguished by checking the error stack when receiving a 0 return value: Because I'm pushing 

```
	RANDerr(RAND_F_RAND_POOL_ADD_END, RAND_R_RANDOM_POOL_OVERFLOW);
```
in the error case.

What wonders me a bit, that you changed the return value of the function (btw: also the return type needs to be changed to int) only to completely ignore the error aftwerwards. Doesn't this prove that a boolean return value is entirely useless here?

Reply:
BTW: the i'm-not-ready-yet-state is also known as the i-want-some-more-state or [oliver-twist-state](https://www.youtube.com/watch?v=gIi7kb3NFyY&feature=youtu.be&t=155) :wink:.

Reply:
> ... because the error state and the i'm-not-ready-yet state are easily distinguished by checking the error stack when receiving a 0 return value: Because I'm pushingbecause the error state and the i'm-not-ready-yet state are easily distinguished by checking the error stack when receiving a 0 return value: Because I'm pushing

Sure, that could be done, but that goes contrary to how we've dealt with this so far.  Either there's a specific value that indicates an error (negative numbers are popular in some APIs), or we go boolean.  In other words, we check the returned value, and *if* it indicates there's an error and are interested in the exact error, we go digging, rather than having to check the error stack to see if the returned value was an error indication or not.
(having to check the errors on the error stack is also not terribly future proof...  consider what happens when a function gets added error pushes with other values, for example)

Regarding the two-liners where the return value of `rand_pool_add` is ignored, it can also be observed that those parts of code already know exactly how much the pool needs to be filled up, so of course we know beforehand that there will be no error (at least until someone changes `rand_pool_add` in such a way that errors *should* be checked here as well).  However, consider some time in the future, someone wanting to do something like this:

``` C
    while ((count = read(bigrandomsource, bigbuffer, sizeof(bigbuffer))) >= 0
           && rand_pool_add(pool, bigbuffer, count, epb * count) == 0)
        ;
```

After all, since `rand_pool_add` supposedly returned non-zero when there is enough entropy, this would be a perfectly intuitive way of using it...  except ambiguity...

With my change, the loop change is actually very small:

``` C
    while (rand_pool_entropy_available(pool) == 0
           && (count = read(bigrandomsource, bigbuffer, sizeof(bigbuffer))) >= 0) {
        if (rand_pool_add(pool, bigbuffer, count, epb * count) == 0)
            goto bail_out;
    }
```

The difference is consistency.

Reply:
Well, i understand your reasoning and I agree that my way of doing it might have advantages but is not consistent with the rest of how OpenSSL is doing it. Now since I'm a great fan of consistency (as you might have noticed), I'll follow your arguments. Approved, provided the return types are changed from `size_t` to `int` (for consistency reasons).